File: src/shapes/SphereAdapter.ts
import * as THREE from 'three';
import type { Cell, Direction } from '../core/Grid';
import type { IShapeAdapter } from './IShapeAdapter';

/**
 * Maps grid coordinates onto a sphere using an equirectangular projection.
 * The sphere uses a single face index (0) for all cells.
 */
export class SphereAdapter implements IShapeAdapter {
  private radius: number;
  constructor(private size = 10, radius?: number) {
    this.radius = radius ?? size / Math.PI;
  }

  getGridSize() {
    return { u: this.size, v: this.size };
  }

  getFaceCount() {
    return 1;
  }

  /**
   * Convert a grid cell to a point on the sphere surface.
   *
   * `u` maps to longitude (phi) and `v` to latitude (theta) using an
   * equirectangular projection. Half-cell offsets center the cell on the
   * surface patch.
   */
  toWorld(cell: Cell): THREE.Vector3 {
    const phi = (cell.u + 0.5) * (2 * Math.PI / this.size); // longitude
    const theta = (cell.v + 0.5) * (Math.PI / this.size); // latitude
    const x = this.radius * Math.sin(theta) * Math.cos(phi);
    const y = this.radius * Math.cos(theta);
    const z = this.radius * Math.sin(theta) * Math.sin(phi);
    return new THREE.Vector3(x, y, z);
  }

  /**
   * Wrap a cell around the sphere.
   *
   * Because the sphere has no edges, moving off one side simply wraps the
   * coordinates modulo the grid size. The direction is unused but kept for
   * parity with other adapters.
   */
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  wrap(cell: Cell, _dir: Direction): Cell {
    let { u, v } = cell;
    u = (u + this.size) % this.size;
    v = (v + this.size) % this.size;
    return { face: 0, u, v };
  }
}

File: python/shapes/sphere_adapter.py
import math
from typing import Tuple

from .ishape_adapter import Cell, IShapeAdapter


class SphereAdapter(IShapeAdapter):
    """Adapter using an equirectangular projection for a sphere surface."""

    def __init__(self, size: int = 10, radius: float | None = None) -> None:
        self.size = size
        self.radius = radius if radius is not None else size / math.pi

    def get_grid_size(self) -> Tuple[int, int]:
        return self.size, self.size

    def get_face_count(self) -> int:
        return 1

    def to_world(self, cell: Cell) -> Tuple[float, float, float]:
        """Convert a grid cell to world coordinates on the sphere."""
        phi = (cell.u + 0.5) * (2 * math.pi / self.size)
        theta = (cell.v + 0.5) * (math.pi / self.size)
        x = self.radius * math.sin(theta) * math.cos(phi)
        y = self.radius * math.cos(theta)
        z = self.radius * math.sin(theta) * math.sin(phi)
        return (x, y, z)

    def wrap(self, cell: Cell, _direction: str) -> Cell:
        """Wrap coordinates around the sphere uniformly."""
        u = (cell.u + self.size) % self.size
        v = (cell.v + self.size) % self.size
        return Cell(0, u, v)
